blockchain/blockvalidator.go

```diff
@@ -67,154 +69,48 @@ func (v *validator) Validate(blk *Block, tipHeight uint64, tipHash hash.Hash32B,
-					return errors.Wrapf(ErrGasHigherThanLimit, "gas is higher than gas limit")
-				}
-				intrinsicGas, err := vote.IntrinsicGas()
-				if intrinsicGas > vote.GasLimit() || err != nil {
-					return errors.Wrapf(ErrInsufficientGas, "insufficient gas for vote")
-				}
-			}
-		case *action.Execution:
-			verifyAction = true
-			execution := act.(*action.Execution)
-			// Verify Address
-			// Verify Nonce
-			// Verify Signature
-			// Verify Gas
-			// Verify Amount
+	producerPKHash := keypair.HashPubKey(blk.Header.Pubkey)
+	producerAddr := address.New(blk.Header.chainID, producerPKHash[:])
+	ctx := protocol.WithValidateActionsCtx(context.Background(),
+		&protocol.ValidateActionsCtx{
+			NonceTracker:    make(map[string]uint64),
+			BlockHeight:     blk.Height(),
+			ProducerAddr:    producerAddr.IotxAddress(),
+			CoinbaseChecked: false,
+		})
 
-			if _, err := iotxaddress.GetPubkeyHash(execution.Executor()); err != nil {
-				return errors.Wrapf(err, "failed to validate executor's address %s", execution.Executor())
-			}
-			if execution.Contract() != action.EmptyAddress {
-				if _, err := iotxaddress.GetPubkeyHash(execution.Contract()); err != nil {
-					return errors.Wrapf(err, "failed to validate contract's address %s", execution.Contract())
+	for _, act := range blk.Actions {
+		vaCtx, _ := protocol.GetValidateActionsCtx(ctx)
+		if act.SrcAddr() != "" {
+			if _, ok := vaCtx.NonceTracker[act.SrcAddr()]; !ok {
+				confirmedNonce, err := v.sf.Nonce(act.SrcAddr())
+				if err != nil {
+					return errors.Wrap(err, "failed to get the confirmed nonce of action sender")
 				}
+				vaCtx.NonceTracker[act.SrcAddr()] = confirmedNonce
 			}
-			// Reject over-gassed execution
-			if execution.GasLimit() > GasLimit {
-				return errors.Wrapf(ErrGasHigherThanLimit, "gas is higher than gas limit")
-			}
-			intrinsicGas, err := execution.IntrinsicGas()
-			if intrinsicGas > execution.GasLimit() || err != nil {
-				return errors.Wrapf(ErrInsufficientGas, "insufficient gas for execution")
-			}
-
-			// Reject execution of negative amount
-			if execution.Amount().Sign() < 0 {
-				return errors.Wrapf(ErrBalance, "negative value")
-			}
-		}
-
-		// Verify signature
-		if verifyAction {
-			wg.Add(1)
-			expectedVerifiedActions++
-			go func(a action.Action, counter *uint64) {
-				defer wg.Done()
-				if err := action.Verify(a); err != nil {
-					return
-				}
-				atomic.AddUint64(counter, uint64(1))
-			}(act, &correctAction)
 		}
-		if verifyNonce {
-			sender := act.SrcAddr()
-			// Store the nonce of the sender and verify later
-			if _, ok := confirmedNonceMap[sender]; !ok {
-				accountNonce, err := v.sf.Nonce(sender)
-				if err != nil {
-					return errors.Wrap(err, "failed to get the nonce of transfer sender")
-				}
-				confirmedNonceMap[sender] = accountNonce
-				accountNonceMap[sender] = make([]uint64, 0)
+		ctx := protocol.WithValidateActionsCtx(context.Background(), vaCtx)
+		for _, validator := range v.actionValidators {
```

Comment:
we should use go func() here to run these in parallel, just like current code?
otherwise i think performance would be hurt a lot